lock-api/src/main/java/com/palantir/lock/watch/ConjureLockRequestMetadataUtils.java

```diff
@@ -0,0 +1,158 @@
+        // visitUnknown() will return null
+        changeMetadata.values().removeIf(Objects::isNull);
+        return LockRequestMetadata.of(changeMetadata);
+    }
+
+    @Unsafe
+    @Value.Immutable
+    public interface ConjureMetadataConversionResult {
+        List<LockDescriptor> lockList();
+
+        ConjureLockRequestMetadata conjureMetadata();
+
+        static ImmutableConjureMetadataConversionResult.Builder builder() {
+            return ImmutableConjureMetadataConversionResult.builder();
+        }
+    }
+
+    private static final class ChangeMetadataToConjureVisitor implements ChangeMetadata.Visitor<ConjureChangeMetadata> {
+        static final ChangeMetadataToConjureVisitor INSTANCE = new ChangeMetadataToConjureVisitor();
+
+        @Override
+        public ConjureChangeMetadata visit(Unchanged unchanged) {
+            return ConjureChangeMetadata.unchanged(ConjureUnchangedChangeMetadata.of());
+        }
+
+        @Override
+        public ConjureChangeMetadata visit(Updated updated) {
+            return ConjureChangeMetadata.updated(
+                    ConjureUpdatedChangeMetadata.of(Bytes.from(updated.oldValue()), Bytes.from(updated.newValue())));
+        }
+
+        @Override
+        public ConjureChangeMetadata visit(Deleted deleted) {
+            return ConjureChangeMetadata.deleted(ConjureDeletedChangeMetadata.of(Bytes.from(deleted.oldValue())));
+        }
+
+        @Override
+        public ConjureChangeMetadata visit(Created created) {
+            return ConjureChangeMetadata.created(ConjureCreatedChangeMetadata.of(Bytes.from(created.newValue())));
+        }
+    }
+
+    private static final class ChangeMetadataFromConjureVisitor
+            implements ConjureChangeMetadata.Visitor<ChangeMetadata> {
+
+        static final ChangeMetadataFromConjureVisitor INSTANCE = new ChangeMetadataFromConjureVisitor();
+
+        @Override
+        public ChangeMetadata visitUnchanged(ConjureUnchangedChangeMetadata unchanged) {
+            return ChangeMetadata.unchanged();
+        }
+
+        @Override
+        public ChangeMetadata visitUpdated(ConjureUpdatedChangeMetadata updated) {
+            return ChangeMetadata.updated(
+                    updated.getOldValue().asNewByteArray(),
+                    updated.getNewValue().asNewByteArray());
+        }
+
+        @Override
+        public ChangeMetadata visitDeleted(ConjureDeletedChangeMetadata deleted) {
+            return ChangeMetadata.deleted(deleted.getOldValue().asNewByteArray());
+        }
+
+        @Override
+        public ChangeMetadata visitCreated(ConjureCreatedChangeMetadata created) {
+            return ChangeMetadata.created(created.getNewValue().asNewByteArray());
+        }
+
+        @Override
+        public ChangeMetadata visitUnknown(String unknownType) {
+            // caller should handle this case
+            return null;
```

Comment:
Failing here is an option, but seems a bit harsh since we might very well want to extend the `ChangeMetadata` interface with additional types in the future.

Reply:
We should use `Optional` here rather than returning `null`. `null` is the source of all evil.

Reply:
Also consider, we do have a subtle invariant here: we can never make decision around absence of information. I think that is OK, but something to be mindful of, as if our interpretation changes overtime, we could really shoot ourselves in the foot.

Reply:
You could also make a `ChangeMetadata` type `Unknown` to handle this case, which is just a singleton. We pay the cost of 4-byte reference + cpu cycles skipping over useless object, but would then give discretion to the implementing developer. I.e., if we see an unknown type, we should fail hard because X/Y/Z. 

(I think Optionals is the best way to go personally)

Reply:
I initially returned `null` here because of performance reasons. Wrapping everything in an optional can be a bit of an overhead, but I am not strictly opposed.